### PR TITLE
5673 unit testing

### PR DESCRIPTION
**Pull request by alexscheitlin**
As mentioned in #5673, I created new unit tests.

For commit `b4ee5b7`, I created a graph (see below) for the following function. The number of the nodes are used to determine prime paths and test paths (see comments in Test Code). Below you will also find the derived test cases.

**Method under Test**
https://github.com/IQSS/dataverse/blob/f3e07a413a7202263ec3c18aa6b9dd556c2b27ba/src/main/java/edu/harvard/iq/dataverse/Dataset.java#L251-L267

**Method Graph**
graph

**Test Cases**

```
[1,2,3,4]
no versions
return false

[1,2,3,5,6]
v1: released
return false

[1,2,3,5,7,8]
v1: draft
return false

[1,2,3,5,7,9,11,3,5,7,8]
v1: not released, not draft, not deaccessioned
v2: draft
return false

[1,2,3,5,7,9,10,11,3,5,6]
v1: deaccessioned
v2: released
return false

[1,2,3,5,7,9,11,3,5,6]
v1: not released, not draft, not deaccessioned
v2: released
return false

[1,2,3,5,7,9,10,11,3,5,7,8]
v1: deaccessioned
v2: draft
return false

[1,2,3,5,7,9,11,3,5,7,9,11,3,4]
v1: not released, not draft, not deaccessioned
v2: not released, not draft, not deaccessioned
return false

[1,2,3,5,7,9,11,3,5,7,9,10,11,3,4]
v1: not released, not draft, not deaccessioned
v2: deaccessioned
return true

[1,2,3,5,7,9,10,11,3,5,7,9,11,3,4]
v1: deaccessioned
v2: not released, not draft, not deaccessioned
return true

[1,2,3,5,7,9,10,11,3,5,7,9,10,11,3,4]
v1: deaccessioned
v2: deaccessioned
return true
```

## Related Issues

- connects to #5673

## Pull Request Checklist

- [x] Unit [tests][] completed
- [x] Integration [tests][]: None
- [x] Deployment requirements, [SQL updates][], [Solr updates][], etc.: None
- [x] [Documentation][docs] completed: None
- [x] Merged latest from "develop" [branch][] and resolved conflicts

[tests]: http://guides.dataverse.org/en/latest/developers/testing.html
[SQL updates]: http://guides.dataverse.org/en/latest/developers/sql-upgrade-scripts.html
[Solr updates]: https://github.com/IQSS/dataverse/blob/develop/conf/solr/7.3.0/schema.xml
[docs]: http://guides.dataverse.org/en/latest/developers/documentation.html
[branch]: http://guides.dataverse.org/en/latest/developers/branching-strategy.html

### CODE REVIEW

**Comment by alexscheitlin** [comment]
> 
> 
> I noticed that the code itself (FileUtil.java) has been modified in this pull request.

Yes, both 3196e02 and caa2def modify the production method so that the introduced test cases are handled properly. In particular, the changes handle inputs beeing `null` or empty Strings.

File: src/main/java/edu/harvard/iq/dataverse/util/FileUtil.java [hunk 1]
```diff
@@ -192,11 +192,17 @@ public static String replaceExtension(String originalName) {
     }   
     
     public static String replaceExtension(String originalName, String newExtension) {
```
Comment (matthew-a-dunlap) [inline src/main/java/edu/harvard/iq/dataverse/util/FileUtil.java]: I understand the urge to make the code more robust, but the changes done to replaceExtension quite possibly have downstream changes. Code may be expecting an error output on null or "" that will no longer be there. Please revert this change.

File: src/test/java/edu/harvard/iq/dataverse/DatasetTest.java [hunk 2]
```diff
@@ -49,5 +56,187 @@ public void testLocksManagement() {
...
+    // 11 test paths (= 11 test methods) are needed.
+    // 4 different dataset versions are needed:
+    // - DEACCESSIONED
+    // - DRAFT
+    // - RELEASED
+    // - non of the above (= ARCHIVED)
+    //
+    // See details here: https://github.com/IQSS/dataverse/pull/5703
+    //
+    // [1] https://cs.gmu.edu:8443/offutt/coverage/GraphCoverage
```
Comment (matthew-a-dunlap) [inline src/test/java/edu/harvard/iq/dataverse/DatasetTest.java]: I left a comment on @nikzaugg 's commit that applies here. Looking at this tool, I do not understand how it was used with our codebase.

See this comment for info on how to proceed: https://github.com/IQSS/dataverse/pull/5705#discussion_r271425094

File: src/test/java/edu/harvard/iq/dataverse/DatasetTest.java [hunk 3]
```diff
@@ -49,5 +56,187 @@ public void testLocksManagement() {
...
+    @After
+    public void after() {
+        this.archivedVersion = null;
+        this.deaccessionedVersion = null;
+        this.draftVersion = null;
+        this.releasedVersion = null;
+    }
+
+    @Test
+    public void testIsDeaccessionedWithNoVersions() {
```
Comment (matthew-a-dunlap) [inline src/test/java/edu/harvard/iq/dataverse/DatasetTest.java]: The title of these tests could be a bit clearer. Maybe "testNotDeaccessionedWithNoVersions". I'm up for suggestion, but these tests imply to me that you are looking for cases where isDeaccessioned is true.

File: src/test/java/edu/harvard/iq/dataverse/util/FileUtilTest.java [hunk 4]
```diff
@@ -61,6 +61,49 @@ public void testGetCiteDataFileFilename() {
...
+                { null, null, "" },
+                { null, null, "y" },
+
+                { "", "", null },
+                { "", "", "" },
+                { "", "", "y" },
+            });
+        }
+
+        @Parameter
```
Comment (matthew-a-dunlap) [inline src/test/java/edu/harvard/iq/dataverse/util/FileUtilTest.java]: could you move these parameter definitions before the Parameters list? its a bit more readable.

File: src/main/java/edu/harvard/iq/dataverse/util/FileUtil.java (hunk 1 above)
Comment (alexscheitlin) [reply src/main/java/edu/harvard/iq/dataverse/util/FileUtil.java]: Hi @matthew-a-dunlap, I understand that. I reverted the changes to the production class in 
3f20267 and a97ed76 and also commented out the test cases that would fail. If you want me to, I will uncomment these test cases and adjust them so that they pass with the current implementation.

File: src/test/java/edu/harvard/iq/dataverse/DatasetTest.java (hunk 2 above)
Comment (alexscheitlin) [reply src/test/java/edu/harvard/iq/dataverse/DatasetTest.java]: I replied to this matter on the PR of @nikzaugg here: https://github.com/IQSS/dataverse/pull/5705#discussion_r271706583

File: src/test/java/edu/harvard/iq/dataverse/DatasetTest.java (hunk 3 above)
Comment (alexscheitlin) [reply src/test/java/edu/harvard/iq/dataverse/DatasetTest.java]: I agree. The wording is misleading. I changed it with commit 82405bb.
Examples are:
- `datasetShouldNotBeDeaccessionedWithNoVersions`
- `datasetShouldBeDeaccessionedWithDeaccessionedAndArchivedVersions`

File: src/test/java/edu/harvard/iq/dataverse/util/FileUtilTest.java (hunk 4 above)
Comment (alexscheitlin) [reply src/test/java/edu/harvard/iq/dataverse/util/FileUtilTest.java]: Sure, I made the change in 2c612b3

File: src/main/java/edu/harvard/iq/dataverse/util/FileUtil.java (hunk 1 above)
Comment (matthew-a-dunlap) [reply src/main/java/edu/harvard/iq/dataverse/util/FileUtil.java]: Yes feel free to get those tests to pass with how things are implemented currently.

File: src/test/java/edu/harvard/iq/dataverse/DatasetTest.java (hunk 2 above)
Comment (matthew-a-dunlap) [reply src/test/java/edu/harvard/iq/dataverse/DatasetTest.java]: Please remove the references to the graphing tool from the tests and write up a basic overview of the cases tested (its fine if it doesn't cover "all the edges of the graph"). I'm glad its been useful to understand and write tests cases, but its too much of a maintenance/code-rot burden to have info about it live on in the code.

File: src/main/java/edu/harvard/iq/dataverse/util/FileUtil.java (hunk 1 above)
Comment (alexscheitlin) [reply src/main/java/edu/harvard/iq/dataverse/util/FileUtil.java]: I adjusted the test cases where it was possible (f9d2001)

File: src/test/java/edu/harvard/iq/dataverse/DatasetTest.java (hunk 2 above)
Comment (alexscheitlin) [reply src/test/java/edu/harvard/iq/dataverse/DatasetTest.java]: I removed the reference to the tool and added a short description in (b378bca).
